Lib/ntpath.py

```diff
@@ -640,23 +638,69 @@ def commonpath(paths):
         genericpath._check_arg_types('commonpath', *paths)
         raise
 
+MAX_PATH = 260
 
-# determine if two files are in fact the same file
 try:
-    # GetFinalPathNameByHandle is available starting with Windows 6.0.
-    # Windows XP and non-Windows OS'es will mock _getfinalpathname.
-    if sys.getwindowsversion()[:2] >= (6, 0):
-        from nt import _getfinalpathname
-    else:
-        raise ImportError
-except (AttributeError, ImportError):
-    # On Windows XP and earlier, two files are the same if their absolute
-    # pathnames are the same.
-    # Non-Windows operating systems fake this method with an XP
-    # approximation.
-    def _getfinalpathname(f):
-        return normcase(abspath(f))
+    from nt import _getfinalpathname
+
+    # determine if two files are in fact the same file
+    def realpath(filename):
+        filename = os.fspath(filename)
+        is_str = isinstance(filename, str)
+        if not is_str:
+            filename = os.fsdecode(filename)
+        extended_path_prefix = '\\\\?\\'
+        is_extended_path = filename.startswith(extended_path_prefix)
+        unresolved = filename if is_extended_path else abspath(filename)
+        resolved_parts = []
+        while True:
+            try:
+                resolved_parts.append(_getfinalpathname(unresolved))
+                break
+            except OSError:
+                unresolved, tail = split(unresolved)
+                if not tail:
+                    resolved_parts.append(unresolved)
+                    break
+                resolved_parts.append(tail)
+        resolved = join(*reversed(resolved_parts))
+        # try to convert extended path to normal if
+        # initial path did not use \\?\ prefix and result uses it
+        if not is_extended_path and resolved.startswith(extended_path_prefix):
+            resolved = _extended_to_normal(resolved)
+        return resolved if is_str else os.fsencode(resolved)
+
+    def _extended_to_normal(path):
+        drive, rest = splitdrive(path)
```

Comment:
What if `splitdrive` gets updated to return "\\\\?\\UNC\server\share" as the drive, as it does for "\\\\server\share"? I wouldn't be against such a change, as a DOS "drive" is not a synonym for a "device". Maybe we should inline a limited version that assumes a device path and splits it into `device` (without the prefix) and `rest`.

Reply:
I'm wondering if we need split at all and if this will be enough:
```
    def _extended_to_normal(path):
        letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
        if len(path) > 6 and path[4].upper() in letters and path[5] == ':':
            # extended path with \\?\ prefix
            # 4 is len('\\?\')
            normal_path = path[4:]
        elif len(path) > 7 and path[:7].upper() == '\\\\?\\UNC':
            # UNC path with \\?\ prefix - drop prefix
            # 7 is len('\\?\UNC')
            normal_path = '\\' + path[7:]
        else:
            # not a UNC or drive-letter path
            # return path as-is
            return path
```

Reply:
I'd be more comfortable if the check includes the root backslash. In principle a device name could be "C:SPAM" or "UNCEGGS". However, it's not critical and shouldn't be a cause for concern since that case should never end up here. If `realpath` is passed a device path that it can't resolve via `_getfinalpathname`, then the prefix shouldn't change. In either case (initially either "\\\\.\\" or "\\\\?\\"), we won't call `_extended_to_normal`. 

Note that `_getfinalpathname` only succeeds for a DOS logical or UNC drive path. We could expand its behavior to fall back on volume GUID and NT names (the latter modified to use the "GlobalRoot" link, e.g. "\\\\?\\GlobalRoot\\Device\\Spam\\Eggs"), but that still wouldn't pose a problem.